### ISSUE TITLE
spurious CR at start of output in quiet mode

### ISSUE THREAD

**Issue by eduardkareletc** [gap-system/gap#430]
Running gap as:
$ echo 'Display("test");; quit;;' |gap -q |od -N 10 -t a
0000000   cr   t   e   s   t  nl 

This shows an initial CR character in the output. In my understanding, the -q flag should have suppressed any output before the output generated by the program  (t e s t nl)". In further output from a gap program each line is terminated by NL, that's what I found inspecting the output of my program that lead to me running the test above to show this bug.

My gap version info is:
 ┌───────┐   GAP, Version 4.7.7 of 13-Feb-2015 (free software, GPL)
 │  GAP                │   http://www.gap-system.org
 └───────┘   Architecture: x86_64-apple-darwin13.4.0-gcc-default64

**Comment by ChrisJefferson** [comment]
I don't seem able to reproduce in 4.7.8, or the current git version. I don't have 4.7.7, could you try 4.7.8 (although I don't think it will fix it...)

```
~/r/g/r/gap4r7p8 $ echo 'Display("test");; quit;;' | bin/gap.sh -q | od -N 10 -t a
0000000    t   e   s   t  nl
0000005
```

**Comment by alex-konovalov** [comment]
This is triggered by some package, because this does not happen with -A option:
```
Mac-0c4de9c1b942:gap4r7p9 alexk$ echo 'Display("test");; quit;;' | bin/gap.sh -q | od -N 10 -t a
0000000   cr   #   I  sp  sp   Y   o   u  sp   m                        
0000012
Mac-0c4de9c1b942:gap4r7p9 alexk$ echo 'Display("test");; quit;;' | bin/gap.sh -q -A -r | od -N 10 -t a
0000000    t   e   s   t  nl                                            
0000005
Mac-0c4de9c1b942:gap4r7p9 alexk$ echo 'Print("test");; quit;;' | bin/gap.sh -q -A -r | od -N 10 -t a
0000000    t   e   s   t                                                
0000004
Mac-0c4de9c1b942:gap4r7p9 alexk$ echo 'Print("test");; quit;;' | bin/gap.sh -q | od -N 10 -t a
0000000   cr   #   I  sp  sp   Y   o   u  sp   m                        
0000012
```

**Comment by alex-konovalov** [comment]
This is triggered by some package, because this does not happen with -A option:
```
$ echo 'Display("test");; quit;;' | bin/gap.sh -q -r| od -N 10 -t a
0000000   cr   t   e   s   t  nl                                        
0000006
$ echo 'Display("test");; quit;;' | bin/gap.sh -q -A -r| od -N 10 -t a
0000000    t   e   s   t  nl                                            
0000005
$ echo 'Print("test");; quit;;' | bin/gap.sh -q -r| od -N 10 -t a
0000000   cr   t   e   s   t                                            
0000005
$ echo 'Print("test");; quit;;' | bin/gap.sh -q -A -r| od -N 10 -t a
0000000    t   e   s   t                                                
0000004
```

**Comment by eduardkareletc** [comment]
Hello Christopher,

The bug no longer exists in 4.7.8, or 4.7.9. I should have checked that 
myself.

Likely the fact that MacOS uses the CR as end-of-line rather than NL mat 
be at the root of this.

Thanks for the help and a happy new year.
Eduard

**Comment by alex-konovalov** [comment]
@eduardkareletc @ChrisJefferson no it is still there and it happens when Browse package by @frankluebeck is loaded - if it's not compiled, you will not be able to reproduce this. See below:
```
$ echo 'Display("test");; quit;;' | bin/gap.sh -q -A -r| od -N 10 -t a
0000000    t   e   s   t  nl                                            
0000005
$ echo 'LoadPackage("browse":OnlyNeeded);;Display("test");; quit;;' | bin/gap.sh -q -A -r| od -N 10 -t a
0000000   cr   t   e   s   t  nl                                        
0000006
```

**Comment by fingolfin** [comment]
> Likely the fact that MacOS uses the CR as end-of-line rather than NL mat be at the root of this.

That's not true at all -- it held for MacOS 9 and older, but Mac OS X behaves like any other Unix in this regard.

**Comment by sebasguts** [comment]
I think it is caused by Browse. If I start GAP with Browse, I get the `\r`, without, I don't.

**Comment by alex-konovalov** [comment]
@sebasguts exactly - see the comment I left 25 days ago.

**Comment by fingolfin** [comment]
This is caused by the `endwin()` call in the `PostRestore()` function of the `Browse` kernel extension. If I comment it out, the spurious CR is gone.

**Comment by frankluebeck** [comment]
I missed this thread so far, sorry.

Indeed, when GAP is started and Browse is installed it will by default load Browse.
In the example code Browse detects that the output is no terminal and so initializes itself for a "dumb" terminal. The CR which is returned upon the `endwin()` call is probably part of the definition of the dumb terminal. 

Please, do not remove this `endwin()` call in PostRestore, it is needed to end the visual mode after initialization of ncurses. (If you start GAP normally, your terminal will be in a bad state - quit GAP and type `reset<CR>` to recover.)

The correct solution is to start GAP without Browse (which is not very useful anyway without an interactive terminal). For example, use the `-A` option and put explicit `LoadPackage(...)` call for the packages you really need into your code.

**Comment by fingolfin** [comment]
Yeah, I didn't mean to suggest that one should remove the `endwin()` call, just wanted to express that I verifie that it is responsible for printing the CR :).

And please feel free to change the label by yourself as you see fit :-).

**Comment by frankluebeck** [comment]
> And please feel free to change the label by yourself as you see fit :-).

Ah, I wasn't aware that even I can understand how do that without reading pages and pages of documentation . . .

**Comment by fingolfin** [comment]
I don't think we'll ever change anything about this. Workarounds were suggested (e.g. use `-A`).

**Comment by hongyi-zhao** [comment]
See [here](https://lists.uni-kl.de/gap/arc/forum/2023-03/msg00083.html) for the related discussion.

**Comment by kiryph** [comment]
The workaround `-A` does not help when packages are used which depend or suggest explicitly or implicitly through other the package `browse`. Packages I have identified so far are:

- `ctbllib`
- `CAP`
- `AtlasRep`
- `Tomlib` (implicitly through atlasrep (needed) and ctbllib (suggested))
- `HeLP` (implicitly through CTblLib and atlasrep)

**Comment by frankluebeck** [comment]
I cannot see a problem here:
```
echo 'Display("test");;LoadPackage("ctbllib");; quit;;' |gap -q -A | od -N 10 -t a
0000000   t   e   s   t  nl  cr
```
Nevertheless, if you want to make sure that under all circumstances `Browse` is not loaded, you can disable it by putting the line
```
SetUserPreference( "PackagesToIgnore", [ "browse" ] );
```
in your `gap.ini` file.
You could also use options `-r` and `-l` for using a specific `gap.ini` file in your scripts (instead of `~/.gap/gap.ini`).

**Comment by kiryph** [comment]
@frankluebeck The order of your command is wrong. You have to print something after loading the package not before:

```
❯ echo 'LoadPackage("ctbllib");;Display("test");; quit;;' |\gap -q -A | od -N 10 -t a
0000000   cr   t   e   s   t  nl
```

**Comment by frankluebeck** [comment]
@kiryph  Yes thanks, this way I can reproduce the `cr`.

(And confirm, that the solution works ...)

**Comment by kiryph** [comment]
@frankluebeck 

Thanks for pointing the user setting out:

> `SetUserPreference( "PackagesToIgnore", [ "browse" ] );`

I was not aware of it and helps to run `gap` even without `-A` and still not loading GAP package `browse`.
